### ISSUE TITLE
Relay server disconnected with: source/source_queue_advance Client has fallen too far behind

### ISSUE THREAD

**Issue by cgrozev** [karlheyes/icecast-kh#96]
Hi! I see that the issue with "fallen-too-far-behind" has been addressed before, but in our case it is happening between the main server and the relay server (both are Icecast 2.3.3-kh10). They are connected via a fiberoptic vlan, so I would not expect network congestion, given that there is only one listener on the main server - and that is the relay server.

We are streaming a total of 15 streams (5 channels in different encoding formats) between the main and the relay server. Today we got 5 disconnects, on various channels/formats, all with identical message in the main server's log:

2014-08-11  11:52:11] INFO source/source_queue_advance Client 89 (xx.xx.xx.xx) has fallen too far behind (3546157675) on /xxxxx, removing
[2014-08-11  11:52:11] INFO source/source_read listener count on /xxxxx now 1

Is this an issue with the latency or sync of the connection between the two servers? Do you think it can be resolved by increasing the queue-size on the main server (currently the queue/timeout settings are as follows:

 <queue-size>524288</queue-size>
  <client-timeout>30</client-timeout>
  <header-timeout>15</header-timeout>
   <source-timeout>10</source-timeout>

Any ideas are very welcome!

Thank you in advance
Christo

**Comment by danielisz** [comment]
Currently I'm facing the same issue on two servers running Debian.
[2015-03-02  14:16:23] INFO source/source_queue_advance Client 234329 (127.0.0.1) has fallen too far behind (9530872408) on /live.mp3, removing
[2015-03-02  14:17:12] INFO source/source_queue_advance Client 233939 (127.0.0.1) has fallen too far behind (9531655408) on /live.mp3, removing
[2015-03-02  14:20:08] INFO source/source_queue_advance Client 234141 (127.0.0.1) has fallen too far behind (9534471597) on /live.mp3, removing

I have roughly 1.000 of these a day. 

I was trying different versions: Icecast 2.3.3-kh8 -> Icecast 2.3.3-kh11, but the same error occurs. Did you get any update with this?

**Comment by karlheyes** [comment]
No I didn't.   There isn't any real difference in the ability of icecast to send content between those versions, if the socket buffers are full then it will lag and eventually your queue size will be reached causing the drop you see.  The possible causes are

the reader (listener or whatever is on 127.0.0.1) is not reading quickly enough.
the CPU is getting saturated to a point where there is just not enough being done (not a likely cause)
You have some unusual limits being applied, socket buffer size being small or something some stall on the networking side of things.

The CPU aspect can be viewed with something like vmstat 1, plenty of idle and cs not going ballistic are generally good signs.
check "netstat -tnp | grep icecast" to see the Recv-Q/Send-Q values, they should be 0 (or close to it) if using localhost
Try to identify the connections that are dropping, see how that application is reacting to it, is it getting enough data through to keep the playback consistent.
Are they any system limitations being reached, something like DDoS could keep certain parts of the system busy starving even localhost connections of service?

karl.

**Comment by danielisz** [comment]
Yep, actually there are packages in Recv-Q/Send-Q. So I checked the CPU and RAM.
The CPU is not saturated and there is no high load. The RAM though is almost used, there is about 50MB free out of 1GB, but 500MB is cached.
Tonight I'm going to add some more RAM to the servers, I'll keep you up to date what is happening

**Comment by karlheyes** [comment]
The lack of free ram is not a big concern if a lot is in cache. The cache gets purged when memory is required.  You could also look at cat /proc/net/sockstat

karl.

**Comment by danielisz** [comment]
This says the following:

$ cat /proc/net/sockstat
sockets: used 1165
TCP: inuse 1005 orphan 58 tw 10 alloc 1006 mem 3417
UDP: inuse 8 mem 2
UDPLITE: inuse 0
RAW: inuse 0
FRAG: inuse 0 memory 0

**Comment by karlheyes** [comment]
Is this still an issue?

karl.

**Comment by danielisz** [comment]
I'm not working with that company anymore but I sent an email to my ex-colleagues to ask the status.

Thank you!
